### PR TITLE
RLOO supports async rewards.

### PR DESCRIPTION
**Pull request by pramodith**
# What does this PR do?
Similar to #4567 but for RLOO.

Fixes # (issue)

## Before submitting
- [X] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [X] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [X] Was this discussed/approved via a GitHub issue? Please add a link
      to it if that's the case.
- [X] Did you make sure to update the documentation with your changes?
- [X] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

### CODE REVIEW

**Comment by pramodith** [comment]
Not a huge fan of the duplication, ideally I'd like to see `_compute_rewards` become a utility or create something like

```
class RLVRTrainer(BaseTrainer):

     def _compute_rewards(..):
         .....

```

Both `GRPOTrainer` and `RLOOTrainer` can then inherit RLVRTrainer and if the implementations of _calculate_rewards diverge in the future they can override it. Let me know if you'd like to move to this approach instead of what I have currently.

**Comment by pramodith** [comment]
Came up with a better approach aided by the coding agents! In the previous approach we create a new event loop every time since `asyncio.run` is called every time `_compute_rewards` is called.

In the new approach we create a new daemon thread if there any async reward functions. A new event loop set to run forever is now created on the daemon thread.

Every-time we need to run the async reward functions they now run on the daemon thread's event loop. This way we avoid any overhead each time `_compute_rewards`

**Comment by qgallouedec** [comment]
> Both `GRPOTrainer` and `RLOOTrainer` can then inherit RLVRTrainer and if the implementations of _calculate_rewards diverge in the future they can override it. Let me know if you'd like to move to this approach instead of what I have currently.

I understand. We’ll keep the duplication approach for now. It makes it easier to move quickly on a single trainer without having to worry about integration constraints or side effects on all other trainers in the library.

Based on past experience with similar abstractions attempt, the long-term tradeoff tends to be unfavorable: abstraction brings relatively little benefit, while adding complexity and making future refactors harder to untangle if requirements diverge.
